### ISSUE TITLE
Error Cannot index number with string "firmwares" while downgrading iPhone 6S to 14.X

### ISSUE THREAD

**Issue by Mikiped00** [edwin170/downr1n#58]
Hello,

I'm trying to downgrade a iPhone6S from 15.4.1 to 14.8. I put it in DFU mode and exec `./downr1n.sh --downgrade 14.3 --debug`, but the next output is given:

```
+ unzip -n ipsw/iPhone_4.7_14.3_18C66_Restore.ipsw -d ipsw/extracted
+ cp -v ipsw/extracted//BuildManifest.plist work/
'ipsw/extracted//BuildManifest.plist' -> 'work/BuildManifest.plist'
+ echo '[*] Got extract the IPSW successfully'
[*] Got extract the IPSW successfully
+ '[' '' = 1 ']'
+ '[' true ']'
+ mkdir -p blobs
+ cd ramdisk
+ chmod +x sshrd.sh
+ echo '[*] Creating ramdisk'
[*] Creating ramdisk
+ ./sshrd.sh 15.6
jq: error (at <stdin>:1): Cannot index number with string "firmwares"
parse error: Invalid numeric literal at line 1, column 9
```
Kind regards

**Comment by edwin170** [comment]
AMD ? it looks like your computer it's not reading the iPhone information, please reboot your pc

**Comment by Mikiped00** [comment]
I'm doing the downgrading in a Ubuntu virtual machine.

**Comment by edwin170** [comment]
> I'm doing the downgrading in a Ubuntu virtual machine.

yeah that's the problem, you can't use virtual machine because it doesn't detect or work well with some tools which dualra1n uses

**Comment by Mikiped00** [comment]
I will try in a local Linux machine, I will keep you in touch with news

**Comment by edwin170** [comment]
> I will try in a local Linux machine, I will keep you in touch with news

okey good luck

**Comment by Mikiped00** [comment]
I got stuck there:

```
4192036 bytes (4,2 MB, 4,0 MiB) copied, 1,17656 s, 3,6 MB/s
img4tool version: 0.197-aca6cf005c94caf135023263cbb5c61a0081804f-RELEASE
Compiled with plist: YES
Found IM4R extracting generator: ok
Saved IM4M to blobs/iPhone8,1-14.3.shsh2
[*] Converting blob
[*] Dumpped SHSH
[*] Patching kernel ...
SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
```

On the iPhone screen, the next message is showing:

`000230.380223 wlan0.[A30] handleAdjustBusy@1551:AjustBusy timeout in 120000ms. busystate 0`

**Comment by edwin170** [comment]
> I got stuck there:
> 
> ```
> [20:27:12.321][3] Connecting to new device on location 0x1001d as ID 1
> [20:27:12.371][3] Connected to v2.0 device 1 on location 0x1001d with serial number ramdisk tool Sep 18 2022 20:14:43
> [*] Mounting filesystems ...
> seputil: Gigalocker file (/mnt7/E737978E-0178-5BB6-9E2D-A26A31F68D02.gl) exists
> seputil: Gigalocker initialization completed
> sepi digest (20 bytes): 1767f6de9a5dc393f43e76f73f8e6b5adba1b8af
> sepi nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> rsep digest (20 bytes): c5606b118223907fcb0aa4099569afc8215757da
> rsep nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> [*] Dumpped SHSH
> [*] Patching kernel
> SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> ```

python3 -m pip install pyimg4 pyliblzfse lzss

**Comment by edwin170** [comment]
> > > I got stuck there:
> > > ```
> > > [20:27:12.321][3] Connecting to new device on location 0x1001d as ID 1
> > > [20:27:12.371][3] Connected to v2.0 device 1 on location 0x1001d with serial number ramdisk tool Sep 18 2022 20:14:43
> > > [*] Mounting filesystems ...
> > > seputil: Gigalocker file (/mnt7/E737978E-0178-5BB6-9E2D-A26A31F68D02.gl) exists
> > > seputil: Gigalocker initialization completed
> > > sepi digest (20 bytes): 1767f6de9a5dc393f43e76f73f8e6b5adba1b8af
> > > sepi nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> > > rsep digest (20 bytes): c5606b118223907fcb0aa4099569afc8215757da
> > > rsep nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> > > [*] Dumpped SHSH
> > > [*] Patching kernel
> > > SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> > > ```
> > 
> > 
> > python3 -m pip install pyimg4 pyliblzfse lzss -U
> 
> The same error is given:
> 
> ```
> img4tool version: 0.197-aca6cf005c94caf135023263cbb5c61a0081804f-RELEASE
> Compiled with plist: YES
> Found IM4R extracting generator: ok
> Saved IM4M to blobs/iPhone8,1-14.8.shsh2
> [*] Converting blob
> [*] Dumpped SHSH
> [*] Patching kernel ...
> SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> ```

https://github.com/dualra1n/dualra1n/issues/74

**Comment by Mikiped00** [comment]
> > > > I got stuck there:
> > > > ```
> > > > [20:27:12.321][3] Connecting to new device on location 0x1001d as ID 1
> > > > [20:27:12.371][3] Connected to v2.0 device 1 on location 0x1001d with serial number ramdisk tool Sep 18 2022 20:14:43
> > > > [*] Mounting filesystems ...
> > > > seputil: Gigalocker file (/mnt7/E737978E-0178-5BB6-9E2D-A26A31F68D02.gl) exists
> > > > seputil: Gigalocker initialization completed
> > > > sepi digest (20 bytes): 1767f6de9a5dc393f43e76f73f8e6b5adba1b8af
> > > > sepi nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> > > > rsep digest (20 bytes): c5606b118223907fcb0aa4099569afc8215757da
> > > > rsep nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> > > > [*] Dumpped SHSH
> > > > [*] Patching kernel
> > > > SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> > > > ```
> > > 
> > > 
> > > python3 -m pip install pyimg4 pyliblzfse lzss -U
> > 
> > 
> > The same error is given:
> > ```
> > img4tool version: 0.197-aca6cf005c94caf135023263cbb5c61a0081804f-RELEASE
> > Compiled with plist: YES
> > Found IM4R extracting generator: ok
> > Saved IM4M to blobs/iPhone8,1-14.8.shsh2
> > [*] Converting blob
> > [*] Dumpped SHSH
> > [*] Patching kernel ...
> > SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> > ```
> 
> [dualra1n/dualra1n#74](https://github.com/dualra1n/dualra1n/issues/74)

Now, the iPhone screen is green and the following error is given:

```
Sending iBEC (436387 bytes)...
[==================================================] 100.0%
./downr1n.sh: line 361: 25257 Segmentation fault      (core dumped) "$dir"/futurerestore -t blobs/"$deviceid"-"$version".shsh2 --use-pwndfu --skip-blob --rdsk work/rdsk.im4p --rkrn work/krnl.im4p --latest-sep "$HasBaseband" "$ipsw"
if you are on linux please try above command if futurerestore failed
if futurerestore didn't finish succesfully please try to run (with sudo or without) this command: /home/iphone/downr1n/binaries/Linux/futurerestore -t blobs/iPhone8,1-14.8.shsh2 --use-pwndfu --skip-blob --rdsk work/rdsk.im4p --rkrn work/krnl.im4p --latest-sep --latest-baseband ipsw/iPhone_4.7_14.8_18H17_Restore.ipsw
if futurerestore restore sucess, you can boot using  --boot
```

**Comment by edwin170** [comment]
> > > > > I got stuck there:
> > > > > ```
> > > > > [20:27:12.321][3] Connecting to new device on location 0x1001d as ID 1
> > > > > [20:27:12.371][3] Connected to v2.0 device 1 on location 0x1001d with serial number ramdisk tool Sep 18 2022 20:14:43
> > > > > [*] Mounting filesystems ...
> > > > > seputil: Gigalocker file (/mnt7/E737978E-0178-5BB6-9E2D-A26A31F68D02.gl) exists
> > > > > seputil: Gigalocker initialization completed
> > > > > sepi digest (20 bytes): 1767f6de9a5dc393f43e76f73f8e6b5adba1b8af
> > > > > sepi nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> > > > > rsep digest (20 bytes): c5606b118223907fcb0aa4099569afc8215757da
> > > > > rsep nonce (20 bytes): b06a849f2d6633edda3fbbed46921f19d9c66f3b
> > > > > [*] Dumpped SHSH
> > > > > [*] Patching kernel
> > > > > SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> > > > > ```
> > > > 
> > > > 
> > > > python3 -m pip install pyimg4 pyliblzfse lzss -U
> > > 
> > > 
> > > The same error is given:
> > > ```
> > > img4tool version: 0.197-aca6cf005c94caf135023263cbb5c61a0081804f-RELEASE
> > > Compiled with plist: YES
> > > Found IM4R extracting generator: ok
> > > Saved IM4M to blobs/iPhone8,1-14.8.shsh2
> > > [*] Converting blob
> > > [*] Dumpped SHSH
> > > [*] Patching kernel ...
> > > SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats
> > > ```
> > 
> > 
> > [dualra1n/dualra1n#74](https://github.com/dualra1n/dualra1n/issues/74)
> 
> Now, the iPhone screen is green and the following error is given:
> 
> ```
> Sending iBEC (436387 bytes)...
> [==================================================] 100.0%
> ./downr1n.sh: line 361: 25257 Segmentation fault      (core dumped) "$dir"/futurerestore -t blobs/"$deviceid"-"$version".shsh2 --use-pwndfu --skip-blob --rdsk work/rdsk.im4p --rkrn work/krnl.im4p --latest-sep "$HasBaseband" "$ipsw"
> if you are on linux please try above command if futurerestore failed
> if futurerestore didn't finish succesfully please try to run (with sudo or without) this command: /home/iphone/downr1n/binaries/Linux/futurerestore -t blobs/iPhone8,1-14.8.shsh2 --use-pwndfu --skip-blob --rdsk work/rdsk.im4p --rkrn work/krnl.im4p --latest-sep --latest-baseband ipsw/iPhone_4.7_14.8_18H17_Restore.ipsw
> if futurerestore restore sucess, you can boot using  --boot
> ```
> 
> I executed the long command, but segmentation fault is given again
> 
> ```
> root@iphone:/home/iphone/downr1n# /home/iphone/downr1n/binaries/Linux/futurerestore -t blobs/iPhone8,1-14.8.shsh2 --use-pwndfu --skip-blob --rdsk work/rdsk.im4p --rkrn work/krnl.im4p --latest-sep --latest-baseband ipsw/iPhone_4.7_14.8_18H17_Restore.ipsw
> Version: v2.0.0-test(9554c0068dc50e141872ced5da2bd95baa595805-308)
> img4tool version: 0.197-aca6cf005c94caf135023263cbb5c61a0081804f-RELEASE
> libipatcher version: 0.88-1e855d70c84419014e363bdbcaead7b145fe3e1f-RELEASE
> Odysseus for 32-bit support: yes
> Odysseus for 64-bit support: yes
> Checking for updates...
> Futurerestore is up to date!
> INFO: device serial number is XXXXX
> [INFO] 64-bit device detected
> futurerestore init done
> reading signing ticket blobs/iPhone8,1-14.8.shsh2 is done
> User specified to use latest signed SEP
> Segmentation fault (core dumped)
> ```

when you use the long command did you use sudo ?

**Comment by Mikiped00** [comment]
> futurerestore

I use it with root user

**Comment by edwin170** [comment]
> > futurerestore
> 
> I use it with root user

what linux ?

**Comment by Mikiped00** [comment]
> > > futurerestore
> > 
> > 
> > I use it with root user
> 
> what linux ?

Ubuntu 22.04

**Comment by Mikiped00** [comment]
> mmh very weird that, if you want try dualra1n, also I just looked that you used 14.3 instead of 14.8 as a argument but it doesn't change much the script

Yes, in the example, I used 14.8 because at the time I paste you the output, I used 14.8 instead of 14.3. The point is that I want to do a downgrade of iOS, could dualra1n do it??

**Comment by edwin170** [comment]
> > mmh very weird that, if you want try dualra1n, also I just looked that you used 14.3 instead of 14.8 as a argument but it doesn't change much the script
> 
> Yes, in the example, I used 14.8 because at the time I paste you the output, I used 14.8 instead of 14.3. The point is that I want to do a downgrade of iOS, could dualra1n do it??

well dualra1n could put an earlier iOS version like iOS 14.8 or 14.0 etc, however this is not a complete downgrade, that is a dual boot so if you have iOS 15.4.1 and you dualboot using dualra1n you will have iOS 15.4.1 and 14.8, to boot iOS 15.4.1 you just would need to reboot and it will boot automatically into iOS 15 and to boot on iOS 14.8 you will need a pc and boot with dualra1n --boot

**Comment by edwin170** [comment]
this will take storage ofc like a normal iOS version

**Comment by Mikiped00** [comment]
> this will take storage ofc like a normal iOS version

Okeey, I will try then. I will tell you any news I get from that problem

**Comment by edwin170** [comment]
alright
